File: C++/ModBurger.cpp
#include "ModBurger.h"

int main(){

	int ngg = pow(2.0, nn);
	double s = 10;

	//initialisation();
	init init_values = initialisation();
	vector<double> opti_G = environment(200, 24, 0.2, 0.7);
	vector<double> newdistr;


	vector<vector<double>> all_distr(opti_G.size()+1, vector<double>(ng, 0));
	all_distr[0] = init_values.gamete_distr;
	cout << "distri" << endl;
	for (int t = 0; t<opti_G.size(); t++){
		newdistr = new_distributions(init_values.gamete_values, init_values.rec_table, all_distr[t], 0.00002, init_values.gamete_scheme, opti_G[t], s);
		all_distr[t+1] = newdistr;
		//cout << "E: "<< opti_G[t] << endl;
		cout << all_distr[t+1][3] << "\t";

	cout << endl;	
	}

	int x;
	cin >> x;

	return 0;
}

//------------------------

// environment will define the optimum genotype value through time

vector<double> environment(int tmax, int period, double d, double L)
{
	opti_G.clear();
	double var = d * L;
	std::normal_distribution<> norm(0, var);
	for (int t = 0; t < tmax; t++)
	{
		opti_G.push_back(norm(rdgen) + 0.5 + L * sin(2 * M_PI * t / period));

	}
	return opti_G;
}

// initialisation will generate the phenotypic values of each loci,
// vector of recombination probabilities and the values of recombination
// probabilities from gamete j/k to gamete i.

init initialisation(void)
{	
	init init_values;
	int s, size, g;
	string bin;
	double sum = 0.0;

	alpha[0] = unif(rdgen);
	sum += alpha[0];

	for (int i = 0; i < n - 1; i++)
	{
		alpha[i + 1] = unif(rdgen); //initialise alleles
		sum += alpha[i + 1];
		r.push_back(0.0);
		r[i] = unif(rdgen) / 2.0; //initialise recombination rate
	}
	// cout << "Allelic values: " << endl;
	for (int i = 0; i < n; i++)
	{
		alpha[i] = 0.5 * alpha[i] / sum; //scale alleles

		// cout << alpha[i] << " ";
	}
	// cout << endl;

	//initialise gametes
	for (int i = 0; i < ng; i++)
	{

		std::bitset<n> alleles(i);
		gametes.push_back(alleles.to_string<char, std::string::traits_type, std::string::allocator_type>());
		sum = 0.0;

		for (int j = 0; j < n; j++)
		{
			if (((int)gametes[i][j]) - 48 == 1)
				sum += alpha[j];
		}
		genotypes.push_back(sum);
	}

	//initiliase recombination table
	vector<vector<vector<double>>> rec_table(ng, (vector<vector<double>>(ng, (vector<double>(ng, 0)))));
	for (int j = 0; j < ng; j++)
	{
		for (int k = 0; k < ng; k++)
		{
			for (int i = 0; i < ng; i++)
			{
				rec_table[j][k][i] = jk_i_recombination(gametes[i], gametes[j], gametes[k], r);
			}
		}
	}

	sum = 0;
	vector<double> distr;
	double x;
	for (int i = 0; i<ng; i++){
		x = unif(rdgen);
		distr.push_back(x);
		sum += x;
	}
	for (int i = 0; i<ng; i++){
		distr[i] /= sum;
	}

	init_values.gamete_values = genotypes;
	init_values.gamete_distr = distr;
	init_values.gamete_scheme = gametes;
	init_values.rec_table = rec_table;
	return init_values;
}

double jk_i_recombination(string i, string j, string k, vector<double> r)
{

	double R = 0.0; // final probability of recombination btw j, k, and i
	double prod = 1.0;
	int ngg = pow(2.0, nn); // number of possible recombinations combinations
	int **t;				// table of all recombinations

	vector<int> rec_vec; // all recombinations possibilities
	vector<string> jk;	 // binary of new j and k

	t = new int *[ngg];
	for (int z = 0; z < ngg; z++)
	{ // z go trough all cmb of recombinations
		rec_vec.clear();
		t[z] = new int[nn];
		std::bitset<nn> allele(z);
		for (int zz = 0; zz < nn; zz++)
		{ // zz go trhough all the rec_pos
			t[z][zz] = allele[zz] * (zz + 1);
			if (t[z][zz] > 0)
				rec_vec.push_back(t[z][zz]);
		}

		if (z > 0)
		{
			jk = permut_gamete(j, k, rec_vec);
		}
		else
		{
			jk.push_back(j);
			jk.push_back(k);
			for (int zz = 1; zz < nn; zz++)
				rec_vec.push_back(-zz);
		}

		for (int a = 0; a < 2; a++)
		{
			if (jk[a] == i)
			{

				int rv = 0;
				prod = 1.0;
				for (int zz = 0; zz < (int)r.size(); zz++)
				{
					if (zz == rec_vec[rv] - 1)
					{
						prod *= r[zz];
						rv++;
					}
					else
						prod *= 1.0 - r[zz];
				}
				R += 0.5 * prod;
			}
		}
	}
	//cout << R << endl;
	return R;
}

vector<string> permut_gamete(string j, string k, vector<int> rec)
{
	int rec_spot;

	string j_bin_rec, k_bin_rec;
	vector<string> jk;

	rec_spot = rec[0];

	for (int z = 0; z < rec_spot; z++)
		j_bin_rec += j[z];
	for (int z = rec_spot; z < (int)j.size(); z++)
		j_bin_rec += k[z];

	for (int z = 0; z < rec_spot; z++)
		k_bin_rec += k[z];
	for (int z = rec_spot; z < (int)k.size(); z++)
		k_bin_rec += j[z];

	rec.erase(rec.begin());

	if ((int)rec.size() > 1)
		permut_gamete(j_bin_rec, k_bin_rec, rec);

	jk.push_back(j_bin_rec);
	jk.push_back(k_bin_rec);

	return jk;
}

// vector<vector<vector<double>>> recombination(vector<double> rec_rate)
// {
// 	vector<vector<vector<double>>> rec_table(ng, (vector<vector<double>>(ng, (vector<double>(ng, 0)))));

// 	for (int i = 0; i<ng; i++){
// 		std::bitset<n> i_bit(i);
// 		string i_bin = i_bit.to_string<char, std::string::traits_type, std::string::allocator_type>();
// 		for (int j = 0; j <ng;  j++){
// 			std::bitset<n> j_bit(j);
// 			string j_bin = j_bit.to_string<char, std::string::traits_type, std::string::allocator_type>();
// 			for (int k = 0; k<ng; k++){
// 				std::bitset<n> k_bit(k);
// 				string k_bin = k_bit.to_string<char, std::string::traits_type, std::string::allocator_type>();
// 				rec_table[i][j][k] = jk_i_recombination(i_bin, j_bin, k_bin, rec_rate);
// 				cout << rec_table[i][j][k] << endl;
// 			}
// 		}
// 	}
// 	return rec_table;
// }

vector<double> new_distributions(vector<double> gamete_values, vector<vector<vector<double>>> dtf_rec,
					   vector<double> gamete_distr, double mut_rate, vector<string> gam_bin, double opti_Gt, double strength)
{
	//cout << opti_Gt << endl
	int ng = pow(2, n);
	vector<double> p_star(ng, 0);
	double mean_fitness = 0;
	for (int j = 0; j < ng; j++)
	{
		for (int k = j; k < ng; k++)
		{
			mean_fitness += exp(-s * gamete_values[j] + gamete_values[k] - pow(opti_Gt, 2)) * gamete_distr[j] * gamete_distr[k];
		}
	}
	// cout << mean_fitness << endl;

	for (int i = 0; i < ng; i++)
	{
		for (int j = 0; j < ng; j++)
		{
			for (int k = j; k < ng; k++)
			{
				p_star[i] += exp(-strength * pow(gamete_values[j] + gamete_values[k] - opti_Gt, 2)) * gamete_distr[j] * gamete_distr[k] * dtf_rec[i][j][k] / mean_fitness;
			}
		}
	}

	double mutation_ij = 0;
	
	for (int i = 0; i < ng; i++)
	{
		double tot_mut = 0;
		for (int j = 0; j < ng; j++)
		{
			
			int nm = 0;
			if (i != j)
			{
				for (int z = 0; z < n; z++)
				{
					if (gam_bin[i][z] != gam_bin[j][z]) nm++;
				}
				mutation_ij = pow(mut_rate, nm);
				tot_mut += tot_mut + p_star[j] * mutation_ij - p_star[i] * mutation_ij;
			}
		}
		p_star[i] += tot_mut;
	}
	return p_star;
}

File: R/mod_burger.R
# Modelling of the paper of Burger 2002 
# study of a range of temporal variation 

# Functions
# To get a binary vector
# Number is a positive integer that will be transform in binary
# n is the number of bit in the vector

# > num_to_bin(23, 4)
# [1] 0 1 1 1

num_to_bin <- function(number, n) {
  bin_vector <- rev(as.numeric(intToBits(number)))
  if(missing(n)) {
    return(bin_vector)
  } else {
    bin_vector[-(1:(length(bin_vector) - n))]
  }
}


# initialisation modelling

# will return  a list with respectively the gamete phenotypic values, the
# initial distribution of gametes, the actual allelic composition of each 
# gamete and the recombination rate between each adjacent pair of loci

genetic_sim <- function(num_loci, haploid = F){
  
  nb_gamete <- 2^num_loci
  
  rec_rate <- runif(num_loci-1, 0, 0.5)
  ifelse(haploid, haploid <- 1, haploid <- 0.5)
  loci <- runif(num_loci, 0, 1)
  
  # scale the loci values to get a heterozygote of phenotypic value 0.5.
  loci <- haploid * loci / sum(loci)
  
  # initialize the gamete values and scheme (allelic composition)
  gamete_values <- rep(0, nb_gamete)
  gamete_scheme <- list(rep(0, nb_gamete))
  
  for (i in 1:(nb_gamete)){
    gamete_scheme[[i]] <- num_to_bin(i-1, n = num_loci)
    gamete_values[i] <- sum(loci[as.logical(num_to_bin(i-1, n = num_loci))])
  }
  gamete_distr <- runif(nb_gamete, 0, 1)
  gamete_distr <- gamete_distr / sum(gamete_distr)
  
  return(list(gamete_values, gamete_distr, gamete_scheme, rec_rate))
}




# Calculation of the optimal phenotype over time

get_optimum <- function(t_max, amplitude, period, stoch_magnitude){
  opti_G <- 1:t_max
  opti_G <- lapply(opti_G,
                   function(t) {rnorm(1, 0.5 + amplitude * sin(2*pi*t/period),
                                              stoch_magnitude * amplitude)})
  return(unlist(opti_G))
}


### recombination probabilities table ###

# will transform two gametes considering a list of recombination position
permut_gamete <- function(j_bin, k_bin, rec_vec){
  
  # Consider the first value of the vector of recombination position
  rec_spot <- rec_vec[1]

  j_bin_rec <- c(j_bin[1:rec_spot], k_bin[(rec_spot+1):length(j_bin)])
  k_bin_rec <- c(k_bin[1:rec_spot], j_bin[(rec_spot+1):length(k_bin)])
  
  if (length(rec_vec) > 1) {
    # Recall the function for the next value of the rec_vec
    permut_gamete(j_bin_rec, k_bin_rec, rec_vec[-1])
  } else {
    return(list(j_bin_rec, k_bin_rec))
  }
}

jk_i_recombination <- function(i_bin, j_bin, k_bin, rec_rate){
  R_jk_to_i <- 0
  n <- length(rec_rate)
  l <- rep(list(0:1), n)
  l <- expand.grid(l)
  l <- t((1:(num_loci-1))*t(l))
  for (t in 1:nrow(l)){
    
    rec_vec = l[t, ]

    if (sum(rec_vec)>0){
      rec_vec <- unlist(lapply(rec_vec, function(x) {x[x!=0]}))
      jk <- permut_gamete(j_bin, k_bin, rec_vec)

    } else {
      jk <- list(j_bin, k_bin)
      rec_vec <- -(1:(num_loci-1))
    }
    if (all(jk[[1]] == i_bin)){
      R_jk_to_i <- R_jk_to_i + 0.5 * prod(rec_rate[rec_vec]) * prod(1-rec_rate[-rec_vec])
    }
    if (all(jk[[2]] == i_bin)){
      R_jk_to_i <- R_jk_to_i + 0.5 * prod(rec_rate[rec_vec]) * prod(1-rec_rate[-rec_vec])
    }
  } 

  return(R_jk_to_i)
}




# Recombination of every cbn

recombination <- function(num_loci, rec_rate){
  
  prob_rec <- matrix(0, ncol = 2^num_loci, nrow = 2^(num_loci*2))
  index <- 0
  nb_gamete <- 2^num_loci
  names_cmb <- NULL
  
  for (j in 1:nb_gamete - 1){
    
    j_bin <- num_to_bin(j, num_loci)
    
    for (k in j:(nb_gamete - 1)){
      
      k_bin <- num_to_bin(k, num_loci)
      names_cmb = c(names_cmb, paste(j+1,',',k+1))
      index <- index + 1
      
      for (i in 1:nb_gamete-1){
        
        i_bin <- num_to_bin(i, num_loci)
        prob_rec[index, i+1] <- jk_i_recombination(i_bin, j_bin, k_bin,
                                                   rec_rate)
      }
    }
  }
  
  prob_rec <- prob_rec[1:length(names_cmb), ]
  dtf_rec <- data.frame(t(prob_rec))
  names(dtf_rec) <- names_cmb
  return(dtf_rec)
}

# Calculation of all pi_star

new_distributions <- function(num_loci, gamete_distr, gamete_values,
                              gamete_scheme, opti_Gt, dtf_rec, mut_rate){
  nb_gamete <- 2^num_loci
  p_star <- rep(0, nb_gamete)

  mean_fitness <- 0
  cool = NULL
  for (j in 1:nb_gamete){
    for (k in j:nb_gamete){

      mean_fitness <- mean_fitness + exp(-strength_selec *
                      (gamete_values[j] + gamete_values[k] - opti_Gt)^2) *
                      gamete_distr[j] * gamete_distr[k]
      cool <- c(cool, gamete_distr[j] * gamete_distr[k])
    }
  }

  for (i in 1:nb_gamete){
    for (j in 1:nb_gamete){
      for (k in j:nb_gamete){
        
        G <- gamete_values[j] + gamete_values[k]
        viability_G <- (exp(-strength_selec * (G - opti_Gt)^2) * gamete_distr[j] *
                          gamete_distr[k] * dtf_rec[i, paste(j,',',k)]) / mean_fitness
        p_star[i] <- p_star[i] + viability_G
      } 
    }
  }

  for (i in 1:nb_gamete) {
    
    tot_mut <- 0
    
    for (j in (1:nb_gamete)[-i]){
      mutation_ij <- mut_rate ^ (sum(gamete_scheme[[i]] != gamete_scheme[[j]]))
      tot_mut <- tot_mut + p_star[j]*mutation_ij - p_star[i]*mutation_ij
    }
    
    gamete_distr[i] <- p_star[i] + tot_mut
  }
  return(list(gamete_distr, mean_fitness))
}

new_distributions_haploid <- function(num_loci, gamete_distr, gamete_values,
                          gamete_scheme, opti_Gt, mut_rate){
  nb_gamete <- 2^num_loci
  p_star <- rep(0, nb_gamete)
  

  for (i in 1:nb_gamete){

      gamete_distr[i] <- exp(-strength_selec * (gamete_values[i] - opti_Gt)^2) * gamete_distr[i]

  }

  
  for (i in 1:nb_gamete) {
    
    tot_mut <- 0
    
    for (j in (1:nb_gamete)[-i]){
      mutation_ij <- mut_rate ^ (sum(gamete_scheme[[i]] != gamete_scheme[[j]]))
      tot_mut <- tot_mut + gamete_distr[j]*mutation_ij - gamete_distr[i]*mutation_ij
    }
    
    gamete_distr[i] <- gamete_distr[i] + tot_mut
  }
  gamete_distr <- gamete_distr / sum(gamete_distr)
  return(list(gamete_distr))
}


